### ISSUE TITLE
Several broken API links in the core howto

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4194]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)| @jesstess reported|
|-|-|
|Trac ID|trac#4194|
|Type|defect|
|Created|2009-12-28 04:53:07Z|
|Branch|https://github.com/twisted/twisted/tree/bad-api-links-4194|

Including but not limited to:

process.xhtml:
 * reactor.spawnProcess()
 * reactor.run()
 * reactor.callWhenRunning()

threading.xhtml:
 * threads.blockingCallFromThread()

<details><summary>Searchable metadata</summary>

```
trac-id__4194 4194
type__defect defect
reporter__jesstess jesstess
priority__normal normal
milestone__ 
branch__branches_bad_api_links_4194 branches/bad-api-links-4194
branch_author__jesstess jesstess
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1261975987000000 1261975987000000
changetime__1263433852000000 1263433852000000
version__None None
owner__ 
cc__jesstess cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_1"></a>|@jesstess commented|
|-|-|

(In [27900]) Branching to 'bad-api-links-4194'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_2"></a>|@jesstess commented|
|-|-|

(In [27901]) Fix several broken API links in the core howto and comment out XXX
todos.

refs [#4194](https://github.com/twisted/twisted/issues/4194).

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_3"></a>|@jesstess removed owner|
|-|-|

Some of the links, eg reactor.run, don't have corresponding API docs that would make for very clear link names and were converted to plain <code></code> blocks. Some had '()' at the end. Many of the PB links were to twisted.spread.flavors classes when the API docs are going to be for twisted.spread.pb.

I also commented out XXX comments. I don't feel strongly about making them go away but I don't think they inspire faith in new users.

It's hard to verify that the new links are correct beyond eyeballing it, so a second pass by someone would be great.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_4"></a>|@khorn commented|
|-|-|

application.html
  - It looks like TimerService actually does have a page in the API docs, so there's no reason to remove this one, unless you want to just for consistency with the other items in the list.

pb-copyable.html
  - Around line 570, it's not clear to me why you removed the API class from
    the link to twisted.spread.flavors.  The page does actually exist, though
    of course pb.Cachable is not listed on the page.
    
    In fact the whole twisted.spread.flavors vs. twisted.spread.pb situation 
    is kind of a mess, documentation-wise. The docs talk about flavors, but
    none of the things being talked about are publicly exposed from that 
    module.
    
    Not sure what (if anything) can/should be done about this.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_5"></a>|@khorn set owner to @jesstess|
|-|-|

Doh, hit submit too soon...

Other than the above, the rest looks good to me.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_6"></a>|@jesstess removed owner|
|-|-|

Thanks for the review, khorn!

The pattern matching got ahead of me - I reverted the removal of links to TimerService in application.xhtml and twisted.spread.flavors in pb-copyable.xhtml.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Replying to [jesstess](#note_3):
> Some of the links, eg reactor.run, don't have corresponding API docs that would make for very clear link names and were converted to plain <code></code> blocks.

Wouldn't [IReactorCore.run](http://twistedmatrix.com/documents/9.0.0/api/twisted.internet.interfaces.IReactorCore.html#run) be the right thing to link to in that case?

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_8"></a>|@jesstess commented|
|-|-|

(In [27919]) Add in a link to twisted.internet.interfaces.IReactorCore.run for
reactor.run in process.xhtml.

refs [#4194](https://github.com/twisted/twisted/issues/4194)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_9"></a>|@jesstess commented|
|-|-|

The reason I'd switched reactor.run to a plain <code> block was because the surrounding material keeps referencing ```reactor.fooMethod``` and the code uses ```reactor.run```, and I thought a link to ```IReactorCore.run``` would be less clear.

How is this as a compromise:

```reactor.<code base="twisted.internet.interfaces.IReactorCore" class="API">run</code>```

so the namespace is consistent but only the "run" is a hyperlink?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

When [#2078](https://github.com/twisted/twisted/issues/2078) was resolved, these docs probably should have been updated.  There is no potential for a race anymore, and no warning will be emitted.  Does that help at all? :)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_11"></a>|@khorn commented|
|-|-|

Replying to [jesstess](#note_9):
> The reason I'd switched reactor.run to a plain <code> block was because the surrounding material keeps referencing ```reactor.fooMethod``` and the code uses ```reactor.run```, and I thought a link to ```IReactorCore.run``` would be less clear.
>
> How is this as a compromise:
>
> ```reactor.<code base="twisted.internet.interfaces.IReactorCore" class="API">run</code>```
>
> so the namespace is consistent but only the "run" is a hyperlink?


IMO, I think this is fine.  It may (will) create a spot we'll have to manually fix up after the Sphinx conversion, but there's already a jillion of those, so one more won't hurt a bit.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra set owner to @jesstess|
|-|-|

Replying to [jesstess](#note_3):
> I also commented out XXX comments. I don't feel strongly about making them go away but I don't think they inspire faith in new users.

Can you open ticket(s) for those comments (if you think it's worth it)?
>
> It's hard to verify that the new links are correct beyond eyeballing it, so a second pass by someone would be great.

I guess lore or the documentation builder doesn't check for broken (API) links. I checked most of them and they seem alright to me, +1 for merge, thanks.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_13"></a>|@jesstess commented|
|-|-|

(In [27987]) Remove reference to spawnProcess/reactor.run race condition.

refs [#4194](https://github.com/twisted/twisted/issues/4194)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_14"></a>|@jesstess commented|
|-|-|

(In [27988]) Remove hidden whitespace.

refs [#4194](https://github.com/twisted/twisted/issues/4194)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_15"></a>|@jesstess set status to `closed`|
|-|-|

(In [27989]) Merge bad-api-links-4194

Author: jesstess
Reviewers: khorn, glyph, exarkun, thijs
Fixes: [#4194](https://github.com/twisted/twisted/issues/4194)

Fix several broken API links in the core howto and comment out XXX todos.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
